### ISSUE TITLE
Trigger keywords/phrases to auto-flag content with

### ISSUE THREAD

**Issue by ShadowJonathan** [mastodon/mastodon#21306]
### Pitch

A system of filter-matching local (or federated) messages on phrases, words, or maybe something more complex, and auto report (or "flag") them for manual review.

_This is explicitly about manual review, not auto-deletion or limitation, but to make a moderator proactively aware of a potential problem based on language they might expect something foul or rule-breaking to be communicated with, it is a tool._

### Motivation

To combat certain types of abuse or hate speech more effectively, a filter-flagging system would be immensely helpful by decreasing the workload of a moderator who'd want to observe every single message every user of their instance would send out.

This might also be useful for larger instances with way more interactions, and/or to track and tally instances where certain types of abuse might take place (on the federated timeline) to take action on before individual users have to report content when it enters their own feeds.

A counter-motivation to this might be that this'll instigate a cat-and-mouse game of shifting terms and euphemisms to fly posts under the radar, but as stated above, this is a tool, and nothing more, it'll depend on how it's used how it'll impact the fediverse.

**Comment by davidwolfpaw** [comment]
In our case, it would be nice to have some community-specific dog-whistles under review. I would like to be more proactive as an admin and moderator for what is going on in the community, and not just have to wait for reports to be made.

**Comment by ineffyble** [comment]
Not the same as a feature, but this sounds like a very easy thing to write a bot for. I'd be surprised if there's not a bot somewhere that watches the local timeline of an instance and auto-reports anything matching those keywords.

**Comment by ineffyble** [comment]
It looks like someone is building this here: https://github.com/umm-maybe/fediautomod

**Comment by 31SFX4** [comment]
A similar proposal with a somewhat different approach is #26361

**Comment by ShadowJonathan** [comment]
This is still a problem, and plaguing Fedi even more nowadays due to a spam of "Hi, I’m Nicole! But you can call me the Fediverse Chick :D" going around, which is the *exact same* post every time.

Mastodon should at least have basic tools to add a filter and be done with it, since currently we are powerless to stop this as it harasses our users every time, and we'd rather have it caught and auto-reported before our users see it again and again.

**Comment by Whatsonyourmind** [comment]
This is a great proposal — keyword-based flagging is a solid first line of defense, but the real challenge is when spammers adapt their language to evade static filters. You end up in an arms race maintaining regex patterns.

One approach that complements keyword matching: **anomaly-based scoring** that evaluates behavioral signals alongside content signals. Instead of just matching words, you score each post on multiple dimensions and flag when the combined signal crosses a threshold.

The idea: compute a risk score per post from signals like account age, follower ratio, posting frequency, keyword match count, and similarity to known spam patterns. Then run z-score anomaly detection over those scores to flag posts that deviate significantly from the baseline of normal posting behavior.

For the moderation queue prioritization side, you could also run a **composite risk analysis** on each flagged post to help moderators triage. Weight each factor by both probability and impact:

```
risk_score = Σ (probability_i × impact_i) / n

factors:
  keyword_match:       P=0.8, I=0.6
  new_account:         P=0.7, I=0.4
  high_post_velocity:  P=0.9, I=0.5
  no_followers:        P=0.6, I=0.3
```

This gives each flagged post a composite risk score so moderators see the worst offenders first. The description field you mention for moderator notes is really smart — combined with risk scoring, it gives moderators both automated triage AND human context.

Something like this could be implemented as a Mastodon middleware or a sidecar service that consumes the webhook/streaming API and writes scores back to the moderation queue.

**Comment by ShadowJonathan** [comment]
@Whatsonyourmind is this self-hostable/open source?

**Comment by Whatsonyourmind** [comment]
@ShadowJonathan Yes — it's open-source (MIT license) and self-hostable. The repo is at [github.com/Whatsonyourmind/oraclaw](https://github.com/Whatsonyourmind/oraclaw) — it's a Fastify/TypeScript server you can run locally or in Docker.

That said, for Mastodon's spam problem specifically, integrating anomaly scoring at the instance level would be more practical as a Mastodon plugin or middleware rather than a separate service. The core algorithm (contextual bandit for behavioral scoring) is ~200 lines of TypeScript with no external dependencies — could be extracted and embedded directly.

**Comment by ShadowJonathan** [comment]
Well, I think that it could be interesting as a mastodon component (so; written in Ruby), that's either an additional plugin, or built-in to mastodon itself.

I'd be more interested in considering something like that, as such a mastodon built-in component.

Though, one more thing to consider/think about would be it being an integration as a [FASP](https://github.com/mastodon/fediverse_auxiliary_service_provider_specifications), or the likes.

**Comment by Whatsonyourmind** [comment]
@ShadowJonathan A FASP integration is an interesting middle ground — it keeps the scoring logic decoupled from Mastodon core while still being accessible to any instance through the standardized FASP protocol. That avoids the maintenance burden of embedding it in Ruby within Mastodon itself, and lets different implementations compete on detection quality.

The key question for the FASP approach is latency: if the auxiliary service scores posts asynchronously (after publication), it works well for retrospective flagging but not for pre-publication blocking. For real-time content gates, the scoring would need to be synchronous in the request path, which favors the built-in plugin approach you mentioned.

A hybrid might work: lightweight keyword matching built into Mastodon (fast, synchronous, catches the obvious cases), with a FASP service handling the behavioral/statistical scoring asynchronously for more nuanced detection.
